libafl_qemu/src/executor.rs

```diff
@@ -50,31 +53,103 @@ pub struct QemuExecutor<'a, C, CM, ED, ET, H, I, OT, S, SM> {
     first_exec: bool,
 }
 
+/// LibAFL QEMU main crash handler.
+/// Be careful, it can come after QEMU's native crash handler if a signal is caught by QEMU but
+/// not by LibAFL.
+///
+/// Signals handlers can be nested.
+///
 /// # Safety
 ///
 /// This should be used as a crash handler, and nothing else.
 #[cfg(feature = "usermode")]
-unsafe fn inproc_qemu_crash_handler<ET, I, S>(
+unsafe fn inproc_qemu_crash_handler<E, EM, ET, I, OF, S, Z>(
     signal: Signal,
     info: &mut siginfo_t,
     mut context: Option<&mut ucontext_t>,
-    _data: &mut InProcessExecutorHandlerData,
+    data: &mut InProcessExecutorHandlerData,
 ) where
     ET: EmulatorModuleTuple<I, S>,
-    I: Unpin,
-    S: Unpin,
+    E: Executor<EM, I, S, Z> + HasObservers,
+    E::Observers: ObserversTuple<I, S>,
+    EM: EventFirer<I, S> + EventRestarter<S>,
+    OF: Feedback<EM, I, E::Observers, S>,
+    S: HasExecutions + HasSolutions<I> + HasCorpus<I> + HasCurrentTestcase<I> + Unpin,
+    Z: HasObjective<Objective = OF>,
+    I: Input + Clone + Unpin,
 {
+    log::debug!("QEMU signal handler has been triggered (signal {signal})");
+
     let puc = match &mut context {
         Some(v) => ptr::from_mut::<ucontext_t>(*v) as *mut c_void,
         None => ptr::null_mut(),
     };
 
-    // run modules' crash callback
-    if let Some(emulator_modules) = EmulatorModules::<ET, I, S>::emulator_modules_mut() {
-        emulator_modules.modules_mut().on_crash_all();
+    if let Some(qemu) = Qemu::get() {
+        // QEMU is already initialized, we have to route the signal to QEMU's handler or
+        // consider it as a host (i.e. fuzzer) signal
+
+        if qemu.is_running() {
+            // QEMU is running, we must determine if we are coming from qemu's signal handler or not
+            log::debug!("Signal has been triggered while QEMU was running");
+
+            match qemu.signal_ctx() {
+                QemuSignalContext::OutOfQemuSignalHandler => {
```

Comment:
I almost forgot what we discussed yesterday...
can you explain more on "In what case, you will end up in this branch?"

for example. if we receive a signal during harness execution, we go to which branch?

Reply:
i guess this is the 1st signal?

Reply:
and later InQemuSignalHandlerTarget is the 2nd signal after qemu delegated signal handling to us right?